File: RizzziGit.EnderBytes.Server/src/Resources/UserRole.cs
using System.Text.Json.Serialization;
using Microsoft.Data.Sqlite;

namespace RizzziGit.EnderBytes.Resources;

using Database;

public enum UserRoleType : byte { Admin, User }

public sealed class UserRoleResource(UserRoleResource.ResourceManager manager, UserRoleResource.ResourceData data) : Resource<UserRoleResource.ResourceManager, UserRoleResource.ResourceData, UserRoleResource>(manager, data)
{
  public new sealed class ResourceManager : Resource<ResourceManager, ResourceData, UserRoleResource>.ResourceManager
  {
    private const string NAME = "UserRole";
    private const int VERSION = 1;

    private const string KEY_USER_ID = "UserID";
    private const string KEY_TYPE = "Type";

    public ResourceManager(MainResourceManager main, Database database) : base(main, database, NAME, VERSION)
    {
      main.Users.ResourceDeleted += (transaction, user) => DbDelete(transaction, new()
      {
        { KEY_USER_ID, ("=", user.Id) }
      });
    }

    protected override ResourceData CreateData(SqliteDataReader reader, long id, long createTime, long updateTime) => new(
      id, createTime, updateTime,
      (long)reader[KEY_USER_ID],
      (UserRoleType)(byte)(long)reader[KEY_TYPE]
    );

    protected override UserRoleResource CreateResource(ResourceData data) => new(this, data);

    protected override void OnInit(DatabaseTransaction transaction, int oldVersion = 0)
    {
      if (oldVersion < 1)
      {
        transaction.ExecuteNonQuery($"alter table {NAME} add column {KEY_USER_ID} integer not null;");
        transaction.ExecuteNonQuery($"alter table {NAME} add column {KEY_TYPE} integer not null;");
      }
    }

    public UserRoleType Get(DatabaseTransaction transaction, UserResource user)
    {
      using SqliteDataReader reader = DbSelect(transaction, new()
      {
        { KEY_USER_ID, ("=", user.Id) }
      }, [], new(1, null), null);

      while (reader.Read())
      {
        return Memory.ResolveFromData(CreateData(reader)).Type;
      }

      return UserRoleType.User;
    }

    public void Set(DatabaseTransaction transaction, UserResource user, UserRoleType type, CancellationToken cancellationToken)
    {
      if (DbUpdate(transaction, new()
      {
        { KEY_TYPE, (byte)type }
      }, new()
      {
        { KEY_USER_ID, ("=", user.Id) }
      }) == 0)
      {
        DbInsert(transaction, new() { { KEY_TYPE, (byte)type } });
      }
    }
  }

  public new sealed record ResourceData(
    long Id,
    long CreateTime,
    long UpdateTime,
    long UserId,
    UserRoleType Type
  ) : Resource<ResourceManager, ResourceData, UserRoleResource>.ResourceData(Id, CreateTime, UpdateTime);

  public long UserId => Data.UserId;
  public UserRoleType Type => Data.Type;
}

File: RizzziGit.EnderBytes.Client/src/lib/server/db/user-role.ts
import type { Knex } from 'knex';
import { DataManager, Database, type Data } from '../db';
import type { User } from './user';

export enum UserRoleType {
  SiteAdmin,
  SiteTechnician
}

export interface UserRole extends Data<UserRoleManager, UserRole> {
  userId: number;
  type: UserRoleType;
}

export class UserRoleManager extends DataManager<UserRoleManager, UserRole> {
  public static readonly NAME = 'UserRole';
  public static readonly VERSION = 1;

  public static readonly KEY_USER_ID = 'userId';
  public static readonly KEY_TYPE = 'type';

  public constructor(db: Database, transaction: () => Knex.Transaction<UserRole>) {
    super(db, transaction, UserRoleManager.NAME, UserRoleManager.VERSION);
  }

  protected async upgrade(table: Knex.AlterTableBuilder, oldVersion: number = 0): Promise<void> {
    if (oldVersion < 1) {
      table.integer(UserRoleManager.KEY_USER_ID).notNullable();
      table.integer(UserRoleManager.KEY_TYPE).notNullable();
    }
  }

  public async list(): Promise<UserRole[]> {
    return await this.db.select('*').from<UserRole>(this.name);
  }

  public async count(): Promise<number> {
    const result = (await this.db.from<UserRole>(this.name).count())[0];

    return result['count(*)'] as number;
  }

  public async create(user: User, type: UserRoleType): Promise<UserRole> {
    return this.insert({
      userId: user.id,
      type
    });
  }

  public deleteAllFromUser(user: User) {
    return this.deleteWhere([[UserRoleManager.KEY_USER_ID, '=', user.id]]);
  }
}

Database.register(UserRoleManager);
